govcd/user_test.go

```diff
@@ -0,0 +1,185 @@
+		{
+			name:       "test_user_vapp_author",
+			roleName:   OrgUserRoleVappAuthor,
+			secondRole: OrgUserRoleVappUser,
+		},
+		{
+			name:       "test_user_vapp_user",
+			roleName:   OrgUserRoleVappUser,
+			secondRole: OrgUserRoleConsoleAccessOnly,
+		},
+		{
+			name:       "test_user_console_access",
+			roleName:   OrgUserRoleConsoleAccessOnly,
+			secondRole: OrgUserRoleCatalogAuthor,
+		},
+		{
+			name:       "test_user_catalog_author",
+			roleName:   OrgUserRoleCatalogAuthor,
+			secondRole: OrgUserRoleOrganizationAdministrator,
+		},
+	}
+
+	for _, ud := range userData {
+		fmt.Printf("# Creating user %s with role %s\n", ud.name, ud.roleName)
+		// Uncomment the following lines to see creation request and response
+		// enableDebugShowRequest()
+		// enableDebugShowResponse()
+		user, err := adminOrg.SimpleCreateUser(OrgUserConfiguration{
+			Name:            ud.name,
+			Password:        "user_pass",
+			RoleName:        ud.roleName,
+			ProviderType:    OrgUserProviderIntegrated,
+			DeployedVmQuota: 10,
+			StoredVmQuota:   10,
+			FullName:        strings.ReplaceAll(ud.name, "_", " "),
+			Description:     "user " + strings.ReplaceAll(ud.name, "_", " "),
+			IsEnabled:       true,
+		})
+		// disableDebugShowRequest()
+		// disableDebugShowResponse()
+		check.Assert(err, IsNil)
+
+		AddToCleanupList(ud.name, "user", user.AdminOrg.AdminOrg.Name, check.TestName())
+		check.Assert(user.User, NotNil)
+		check.Assert(user.User.Name, Equals, ud.name)
+		check.Assert(user.GetRoleName(), Equals, ud.roleName)
+		check.Assert(user.User.IsEnabled, Equals, true)
+
+		err = user.Disable()
+		check.Assert(err, IsNil)
+		check.Assert(user.User.IsEnabled, Equals, false)
+
+		fmt.Printf("# Updating user %s with role %s\n", ud.name, ud.secondRole)
+		err = user.ChangeRole(ud.secondRole)
+		check.Assert(err, IsNil)
+		check.Assert(user.GetRoleName(), Equals, ud.secondRole)
+
+		err = user.Enable()
+		check.Assert(err, IsNil)
+		check.Assert(user.User.IsEnabled, Equals, true)
+		err = user.ChangePassword("new_pass")
+		check.Assert(err, IsNil)
+	}
+
+	var enableMap = map[bool]string{
+		true:  "enabled",
+		false: "disabled",
+	}
+	for _, ud := range userData {
+		user, err := adminOrg.GetUserByNameOrId(ud.name, true)
+		check.Assert(err, IsNil)
+
+		fmt.Printf("# deleting user %s (%s - %s)\n", ud.name, user.GetRoleName(), enableMap[user.User.IsEnabled])
+		// uncomment the following two lines to see the deletion request and response
+		// enableDebugShowRequest()
+		// enableDebugShowResponse()
+		err = user.SafeDelete()
+		// disableDebugShowRequest()
+		// disableDebugShowResponse()
+		check.Assert(err, IsNil)
+		user, err = adminOrg.GetUserByNameOrId(user.User.ID, true)
+		check.Assert(err, NotNil)
+		// Tests both the error directly and the function IsNotFound
+		check.Assert(err, Equals, ErrorEntityNotFound)
+		check.Assert(IsNotFound(err), Equals, true)
+		// Expect a null pointer when user is not founc
+		check.Assert(user, IsNil)
+	}
+}
```

Comment:
As far as I guess it's hard to write test for others API methods like `TakeOwnership` ? or it's just not yet done?

Reply:
It is time consuming, as we need running vApps to prove that it does what we assume.
I proved it with manual testing so far. 
We may use some of the existing vApp tests to prove this point, but it is not straightforward.
A practical difficulty would be that we need to either create the vApps using a Org User different from the one running the tests, or assign ownership of some objects to a different user (for which we don't have the functionality in the library yet).

Reply:
As I expected :) So let's add TODO with explanation why we didn't done yet :) -> for someone who forgets it's complicated

Reply:
done